tsl/src/nodes/compressed_skip_scan/exec.c

```diff
@@ -0,0 +1,393 @@
+	Plan *compressed_scan = linitial(cscan->custom_plans);
+	Assert(list_length(cscan->custom_plans) == 1);
+
+	PlanState *ps = &node->ss.ps;
+	if (ps->ps_ProjInfo)
+	{
+		/*
+		 * if we are projecting we need to constify tableoid references here
+		 * because decompressed tuple are virtual tuples and don't have
+		 * system columns.
+		 *
+		 * We do the constify in executor because even after plan creation
+		 * our targetlist might still get modified by parent nodes pushing
+		 * down targetlist.
+		 */
+		List *tlist = ps->plan->targetlist;
+		List *modified_tlist = constify_tableoid(tlist, cscan->scan.scanrelid, state->chunk_relid);
+
+		if (modified_tlist != tlist)
+		{
+			ps->ps_ProjInfo =
+				ExecBuildProjectionInfo(modified_tlist,
+										ps->ps_ExprContext,
+										ps->ps_ResultTupleSlot,
+										ps,
+										node->ss.ss_ScanTupleSlot->tts_tupleDescriptor);
+		}
+	}
+
+	state->hypertable_compression_info = ts_hypertable_compression_get(state->hypertable_id);
+
+	initialize_column_state(state);
+
+	node->custom_ps = lappend(node->custom_ps, ExecInitNode(compressed_scan, estate, eflags));
+
+	state->per_batch_context = AllocSetContextCreate(CurrentMemoryContext,
+													 "DecompressChunk per_batch",
+													 ALLOCSET_DEFAULT_SIZES);
+}
+
+static void
+initialize_batch(CompressedSkipScanState *state, TupleTableSlot *slot)
+{
+	Datum value;
+	bool isnull;
+	int i;
+	MemoryContext old_context = MemoryContextSwitchTo(state->per_batch_context);
+	MemoryContextReset(state->per_batch_context);
+
+	for (i = 0; i < state->num_columns; i++)
+	{
+		CompressedChunkColumnState *column = &state->columns[i];
+
+		switch (column->type)
+		{
+			case COMPRESSED_COLUMN:
+			{
+				value = slot_getattr(slot, column->compressed_scan_attno, &isnull);
+				if (!isnull)
+				{
+					CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
+
+					column->compressed.iterator =
+						tsl_get_decompression_iterator_init(header->compression_algorithm,
+
+															state->reverse)(PointerGetDatum(header),
+
+																			column->typid);
+				}
+				else
+					column->compressed.iterator = NULL;
+
+				break;
+			}
+			case SEGMENTBY_COLUMN:
+				value = slot_getattr(slot, column->compressed_scan_attno, &isnull);
+				if (!isnull)
+					column->segmentby.value = value;
+				else
+					column->segmentby.value = (Datum) 0;
+
+				column->segmentby.isnull = isnull;
+				break;
+			case COUNT_COLUMN:
+				value = slot_getattr(slot, column->compressed_scan_attno, &isnull);
+				state->counter = DatumGetInt32(value);
+				/* count column should never be NULL */
+				Assert(!isnull);
+				break;
+			case SEQUENCE_NUM_COLUMN:
+				/*
+				 * nothing to do here for sequence number
+				 * we only needed this for sorting in node below
+				 */
+				break;
+		}
+	}
+	state->initialized = true;
+	MemoryContextSwitchTo(old_context);
+}
+
+static TupleTableSlot *
+compressed_skip_scan_exec(CustomScanState *node)
```

Comment:
These functions are not different from `DecompressChunk` functions, right? Can you please remove these copies and use the `DecompressChunk` functions directly instead?